src/TesApi.Web/BatchScheduler.cs

```diff
@@ -285,6 +292,92 @@ private bool TryGetCromwellTmpFilePath(string fileUri, out string localPath)
+            var displayName = $"{vmName}:{vmSize}:{isPreemptable}:{registryServer}:{identityResourceId}";
+            var hash = ConvertToBase32(SHA1.HashData(Encoding.UTF8.GetBytes(displayName))).TrimEnd('='); // This becomes 32 chars
+
+            // Build a PoolName that is of legal length, while exposing the most important metadata without requiring user to find DisplayName
+            // Note that the hash covers all necessary parts to make name unique, so limiting the size of the other parts is not expected to appreciably change the risk of collisions. Those other parts are for convenience
+            var remainingLength = PoolKeyLength - hash.Length - 2; // 50 is max name length, 2 is number of inserted chars. This will always be 16 if we use an entire SHA1
+            var visibleVmSize = LimitVmSize(vmSize, Math.Max(remainingLength - vmName.Length, 6));
+            var visibleHostName = vmName[0..Math.Min(vmName.Length, remainingLength - visibleVmSize.Length)];
+            var name = LimitChars($"{visibleHostName}-{visibleVmSize}-{hash}");
+
+            // Trim DisplayName if needed
+            if (displayName.Length > 1024)
+            {
+                // Remove "path" of identityResourceId
+                displayName = displayName[..^identityResourceId.Length] + identityResourceId[(identityResourceId.LastIndexOf('/') + 1)..];
+                if (displayName.Length > 1024)
+                {
+                    // Trim end, leaving fake elipsys as marker
+                    displayName = displayName[..1021] + "...";
+                }
+            }
+
+            return (name, displayName);
+
+            static string LimitVmSize(string vmSize, int limit)
+            {
+                // First try optimizing by removing "Standard_" prefix.
+                var standard = "Standard_";
+                return vmSize.Length <= limit
+                    ? vmSize
+                    : vmSize.StartsWith(standard, StringComparison.OrdinalIgnoreCase)
+                    ? LimitVmSize(vmSize[standard.Length..], limit)
+                    : vmSize[^limit..];
+            }
+
+            static string LimitChars(string text) // ^[a-zA-Z0-9_-]+$
+            {
+                return new(text.AsEnumerable().Select(Limit).ToArray());
+
+                static char Limit(char ch)
+                    => ch switch
+                    {
+                        var x when x >= '0' || x <= '9' => x,
+                        var x when x >= 'A' || x <= 'Z' => x,
+                        var x when x >= 'a' || x <= 'z' => x,
+                        '_' => ch,
+                        '-' => ch,
+                        _ => '_',
+                    };
+            }
+        }
+
+        private static string ConvertToBase32(byte[] bytes) // https://datatracker.ietf.org/doc/html/rfc4648#section-6
+        {
+            const int groupBitlength = 5;
+            var Rfc4648Base32 = @"ABCDEFGHIJKLMNOPQRSTUVWXYZ234567".ToArray();
```

Comment:
Yes, we could. Right now, it's only used in this one place. If I pull it out, I'd probably move it to `BatchUtils` for want of a better home.

Reply:
My comment about pulling it out was in regards to easy performance gains

Reply:
True, converting a constant string into an array of char for each TesTask we turn into a CloudJob (+ once for each created CloudPool) is a waste of computational power, but compared to everything else, it'd take some time before you'd notice its existence. Moving the declaration away from the code that consumes it affects readability more than the performance gains, IMHO.

I could change it into an array of char directly in code...

Reply:
It's the conversion, and also the memory allocation and garbage collection each time the method is run.  Personally I think moving it away from the declaration is fine since "ConvertToBase32" should be self-explanatory; if someone wants to read its implementation, they can do the extra work to find the variables it uses